Fix: normalize punctuation when computing CleanName so searches without punctuation match (closes #1674)

audrey-inglish opened the pull request:
**Summary**
Normalize punctuation/symbols to spaces when computing CleanName (storage-time). This makes searches like "mr robot" match a video titled "Mr. Robot," as described in issue #1674. 

**Changes**
GetCleanValue now delegates to the new NormalizeNameForSearch helper method. This method treats punctuation and symbols (anything that isn't a letter or digit) as whitespace. Then, it strips out excess whitespace. 

I was able to reproduce the issue locally and successfully fail unit tests to demonstrate the behavior. After implementing the change and rescanning the library, the tests pass locally and the search behaves as expected.

**Issues**
Fixes #1674

JPVenson commented on Jellyfin.Server.Implementations/Item/BaseItemRepository.cs:
```diff
@@ -1362,7 +1362,37 @@ private string GetCleanValue(string value)
+        return NormalizeNameForSearch(value);
+    }
+
+    private static string NormalizeNameForSearch(string value)
```
i dont see a reason for adding this as its own dedicated method.
Just include that in the normalise method

theguymadmax commented on Jellyfin.Server.Implementations/Item/BaseItemRepository.cs:
```diff
@@ -1362,7 +1362,37 @@ private string GetCleanValue(string value)
+        }
+
+        // Collapse multiple spaces into single space and trim.
+        var collapsed = System.Text.RegularExpressions.Regex.Replace(sb.ToString(), "\\s+", " ").Trim();
```
You could drop the Regex.Replace and handle space collapsing directly in the loop using a flag.

audrey-inglish replied:
Good idea, thanks! I updated the loop to match what I think you're describing.

theguymadmax commented:
@

theguymadmax commented:
@audrey-inglish Your PR has already been approved, so any changes would need to go through re-approval. I believe those suggestions came from an AI bot. I ran a comparison test on a small library, and the updated version takes about 2 minutes longer than the approved one. When scaled, this could result in a noticeable performance impact. I recommend reverting to the approved version.

Also, please note that there is currently a freeze on any new enhancements until the 10.11 release. As such, it's unlikely that this will be merged before then.

Shadowghost commented:
One question: Does this require a scan with replace all metadata? If that's the case a proper migration might be warranted to do this for existing items once

theguymadmax commented:
> One question: Does this require a scan with replace all metadata? If that's the case a proper migration might be warranted to do this for existing items once Just out of interest since I ran into it recently, how would V/H/S be searchable?

I still had this PR open so i tested it. A scan for missing metadata is required to generate an updated cleanname. Also, no, this doesn't handle "/" in titles. Another example, "Face/Off"

Shadowghost commented:
Ideally we'd handle this case too and then have a migration to fix it.

audrey-inglish commented:
@theguymadmax Thanks for explaining that, and thanks for testing out the performance. I have reverted back to the approved commit.

As for the notes from @Shadowghost , I've started looking into those changes. Handling slashes is already covered in the approved commit I reverted to. I added tests to verify the behavior. Searching for "face off" will yield a show titled "Face/Off."

As far as running a proper migration, I'm unsure on how to handle that correctly. Any direction you can give on this would be appreciated.

Thanks!

JPVenson commented on Jellyfin.Server/Migrations/Routines/RefreshCleanNames.cs:
```diff
@@ -0,0 +1,105 @@
+            try
+            {
+                var newCleanName = BaseItemRepository.GetCleanValue(item.Name ?? string.Empty);
+                if (newCleanName != item.CleanName)
```
```suggestion
                if (!string.Equals(newCleanName, item.CleanName, StringComparison.Ordinal))
```

JPVenson commented on Jellyfin.Server/Migrations/Routines/RefreshCleanNames.cs:
```diff
@@ -0,0 +1,105 @@
+        {
+            try
+            {
+                var newCleanName = BaseItemRepository.GetCleanValue(item.Name ?? string.Empty);
```
```suggestion
                var newCleanName = string.IsNullOrWhiteSpace(item.Name) ? string.Empty : BaseItemRepository.GetCleanValue(item.Name);
```
